### PR TITLE
[SPARK-27555][SQL]HiveSerDe should fall back to hadoopconf if  hive.default.fileformat is not found in SQLConf

### PR DESCRIPTION
**Pull request by sandeep-katta**
## What changes were proposed in this pull request?

SQLConf does not load hive-site.xml.So HiveSerDe should fall back to hadoopconf if  hive.default.fileformat is not found in SQLConf

## How was this patch tested?

Tested manually.
Added UT

### CODE REVIEW

**Comment by sandeep-katta** [comment]
cc @srowen  @HyukjinKwon @cloud-fan

**Comment by srowen** [comment]
I really don't know this part well. Should Spark really need Hive's config directly? this seems a bit specific to certain types of deployment contexts.

**Comment by sandeep-katta** [comment]
may be, but current issue is session state conf(sqlConf) does not load hadoop and hive configurations, so this PR will fix this issue, I see no impact if sqlConf reads all these configurations.

**Comment by cloud-fan** [comment]
I think it's done in `SharedState`?

**Comment by wanghui93** [comment]
I think  sqlConf should load at least hive-site.xml since spark  need 'hive.default.fileformat' which comes from hive-site.xml.

**Comment by wanghui93** [comment]
> I think it's done in `SharedState`?

The configuration parameters in hive-site.xml are loaded into session.sparkContext.hadoopConfiguration parameters by SharedState, not copy to session.sparkContext.conf. And sqlConf just copy session.sparkContext.conf.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 1]
```diff
@@ -152,4 +152,9 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
       session.sparkContext.hadoopConfiguration.unset(mySpecialKey)
     }
   }
+
+  test("SPARK-27555: Spark SessionState.conf should load hive-site.xml ") {
+    val session = SparkSession.builder().master("local").getOrCreate()
+    assert(session.sessionState.conf.getConfString("hive.in.test") == "true")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala]: I'm not sure we want to do this. We load `hive-site.xml` into hadoop conf for hive compatibility. But `hive-site.xml` is not a native/official way to load Spark configs, so this conf should not appear in `SessionStage.conf`.

Please use the `spark-defaults.conf`.

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala]: I'm not sure we want to do this. We load `hive-site.xml` into hadoop conf for hive compatibility. But `hive-site.xml` is not a native/official way to load Spark configs, so this conf should not appear in `SessionStage.conf`.

Please use the `spark-defaults.conf`.

Comment (sandeep-katta) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala]: "hive.in.test" is present in hive-site.xml which is loaded into hadoop conf, so this test asserts presence of hive conf in sessionState.conf, more over we cannot use hive conf inside spark-defaults.conf

**Comment by cloud-fan** [comment]
shall we just make `HiveSerDe.getDefaultStorage` accept hadoop conf?

**Comment by sandeep-katta** [comment]
no we can't, reason is any runtime conf set using set command will not be available in hadoopconf, it is only present in sqlConf

**Comment by HyukjinKwon** [comment]
> shall we just make `HiveSerDe.getDefaultStorage` accept hadoop conf?

Yea, I was about to say that. Why don't we use fall back to Hadoop conf if that's not set in SQLConf?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala [hunk 2]
```diff
@@ -88,6 +103,7 @@ abstract class BaseSessionStateBuilder(
     parentState.map(_.conf.clone()).getOrElse {
       val conf = new SQLConf
       mergeSparkConf(conf, session.sparkContext.conf)
+      mergeHadoopConf(conf, session.sparkContext.hadoopConfiguration)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: This looks basically we're going to mix up Spark conf and Hadoop conf

**Comment by sandeep-katta** [comment]
but SparkSqlParser does not have hadoopConf, it only has sqlConf, so is there any way to get hadoopConf rather than adding new constructor ?

**Comment by HyukjinKwon** [comment]
@sandeep-katta, how about https://github.com/sandeep-katta/spark/pull/1 way? You can merge that PR to your branch so that the commit is added into your PR here.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeSuite.scala [hunk 3]
```diff
@@ -20,14 +20,14 @@ package org.apache.spark.sql.hive.execution
 import java.net.URI
 
 import org.scalatest.BeforeAndAfterAll
-
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeSuite.scala]: Can you revert this unnecessary change?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeSuite.scala [hunk 4]
```diff
@@ -210,4 +210,14 @@ class HiveSerDeSuite extends HiveComparisonTest with PlanTest with BeforeAndAfte
     val e8 = intercept[IllegalArgumentException](analyzeCreateTable(v8))
     assert(e8.getMessage.contains("invalid fileFormat: 'wrong'"))
   }
+
+  test("SPARK-27555: fall back to HDFS if hive.default.fileformat is not found in SQLConf ") {
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeSuite.scala]: fall back to HDFS -> fall back to hive-site.xml?

**Comment by HyukjinKwon** [comment]
ok to test

**Review by viirya — COMMENTED** [review]
The current change looks more reasonable. It is a behavior change by some means, though not big one. Should we update migration guide?

**Review by HyukjinKwon — APPROVED** [review]
Yea, let's do that. Looks good otherwise

File: docs/sql-migration-guide-upgrade.md [hunk 5]
```diff
@@ -126,6 +126,8 @@ license: |
 
   - Since Spark 3.0, parquet logical type `TIMESTAMP_MICROS` is used by default while saving `TIMESTAMP` columns. In Spark version 2.4 and earlier, `TIMESTAMP` columns are saved as `INT96` in parquet files. To set `INT96` to `spark.sql.parquet.outputTimestampType` restores the previous behavior.
 
+  - Since Spark 3.0, if `hive.default.fileformat` is not found in sqlConf then it will fallback to hive-site.xml present in the `hadoopConfiguration` of 'SparkContext'.
```
Comment (HyukjinKwon) [inline docs/sql-migration-guide-upgrade.md]: `sqlConf` -> Spark SQL configuration

Comment (HyukjinKwon) [inline docs/sql-migration-guide-upgrade.md]: `hadoopConfiguration` -> `Hadoop configuration`.

**Review by HyukjinKwon — APPROVED** [review]

File: docs/sql-migration-guide-upgrade.md [hunk 6]
```diff
@@ -126,6 +126,8 @@ license: |
 
   - Since Spark 3.0, parquet logical type `TIMESTAMP_MICROS` is used by default while saving `TIMESTAMP` columns. In Spark version 2.4 and earlier, `TIMESTAMP` columns are saved as `INT96` in parquet files. To set `INT96` to `spark.sql.parquet.outputTimestampType` restores the previous behavior.
 
+  - Since Spark 3.0, if `hive.default.fileformat` is not found in sqlConf then it will fallback to hive-site.xml present in the `Hadoop configuration` of 'SparkContext'.
```
Comment (viirya) [inline docs/sql-migration-guide-upgrade.md]: 'SparkContext' -> ``` `SparkContext` ```

Comment (viirya) [inline docs/sql-migration-guide-upgrade.md]: `sqlConf` -> `Spark SQL configuration`, as @HyukjinKwon suggested.

**Review by viirya — APPROVED** [review]
Two points in the doc. Otherwise looks good.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala [hunk 7]
```diff
@@ -88,7 +89,16 @@ object HiveSerDe {
   }
 
   def getDefaultStorage(conf: SQLConf): CatalogStorageFormat = {
-    val defaultStorageType = conf.getConfString("hive.default.fileformat", "textfile")
+    // To respect hive-site.xml, it peeks Hadoop configuration from existing Spark session,
```
Comment (sujith71955) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala]: Small correction is required in comments
Nit: it picks

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala]: oh, I thought `peeks` makes sense too. It takes a look for Hadoop configuration :).

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeSuite.scala [hunk 8]
```diff
@@ -210,4 +211,23 @@ class HiveSerDeSuite extends HiveComparisonTest with PlanTest with BeforeAndAfte
...
+      val sqlConf = new SQLConf()
+      var storageFormat = HiveSerDe.getDefaultStorage(sqlConf)
+      assert(storageFormat.serde.
+        contains("org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"))
+      // should take orc as it is present in sqlConf
+      sqlConf.setConfString("hive.default.fileformat", "orc")
+      storageFormat = HiveSerDe.getDefaultStorage(sqlConf)
+      assert(storageFormat.serde.contains("org.apache.hadoop.hive.ql.io.orc.OrcSerde"))
+    }
+    finally {
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeSuite.scala]: nit: we could make this line up.

File: docs/sql-migration-guide-upgrade.md [hunk 9]
```diff
@@ -126,6 +126,8 @@ license: |
 
   - Since Spark 3.0, parquet logical type `TIMESTAMP_MICROS` is used by default while saving `TIMESTAMP` columns. In Spark version 2.4 and earlier, `TIMESTAMP` columns are saved as `INT96` in parquet files. To set `INT96` to `spark.sql.parquet.outputTimestampType` restores the previous behavior.
 
+  - Since Spark 3.0, if `hive.default.fileformat` is not found in `Spark SQL configuration` then it will fallback to hive-site.xml present in the `Hadoop configuration` of `SparkContext`.
```
Comment (HyukjinKwon) [inline docs/sql-migration-guide-upgrade.md]: I meant "Spark SQL configuration" without backquotes (not `` `Spark SQL configuration` `` ) but that;s fine.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala (hunk 7 above)
Comment (sujith71955) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala]: yeah you are right, it makes sense too :)
